webauthn/Sources/App/Controllers/WebAuthnController.swift

```diff
@@ -20,167 +20,126 @@ import HummingbirdCore
             challenge: options.challenge
-        )
+        ))
         try await request.session.save(session: session, expiresIn: .minutes(10))
         return options
     }
 
     /// End Authenticating a user
-    struct FinishAuthenticationHandler: HBAsyncRouteHandler {
-        typealias Input = AuthenticationCredential
-        typealias Output = HTTPResponseStatus
-
-        let input: AuthenticationCredential
-        let authenticationSession: AuthenticationSession
-
-        init(from request: HBRequest) throws {
-            self.authenticationSession = try request.authRequire(AuthenticationSession.self)
-            self.input = try request.decode(as: AuthenticationCredential.self)
+    func finishAuthentication(request: HBRequest) async throws -> HTTPResponseStatus {
+        let authenticationSession = try request.authRequire(AuthenticationSession.self)
+        let input = try request.decode(as: AuthenticationCredential.self)
+        guard case .authenticating(let challenge) = authenticationSession else { throw HBHTTPError(.unauthorized) }
+        let id = input.id.urlDecoded.asString()
+        guard let webAuthnCredential = try await WebAuthnCredential.query(on: request.db)
+            .filter(\.$id == id)
+            .with(\.$user)
+            .first()
+        else {
+            throw HBHTTPError(.unauthorized)
         }
-
-        func handle(request: HBRequest) async throws -> Output {
-            guard case .authenticating(let challenge) = self.authenticationSession else { throw HBHTTPError(.unauthorized) }
-            guard let webAuthnCredential = try await WebAuthnCredential.query(on: request.db)
-                .filter(\.$id == input.id.asString())
-                .with(\.$user)
-                .first()
-            else {
-                throw HBHTTPError(.unauthorized)
-            }
-            request.logger.info("Challenge: \(challenge)")
-            do {
-                _ = try request.webauthn.finishAuthentication(
-                    credential: self.input,
-                    expectedChallenge: challenge.urlEncoded,
-                    credentialPublicKey: [UInt8](webAuthnCredential.publicKey.decoded!),
-                    credentialCurrentSignCount: 0
-                )
-            } catch {
-                request.logger.error("\(error)")
-                throw HBHTTPError(.unauthorized)
-            }
-            let session = WebAuthnSessionAuthenticator.Session.authenticated(userId: webAuthnCredential.user.id!)
-            try await request.session.save(session: session, expiresIn: .hours(24))
-
-            return .ok
+        request.logger.info("Challenge: \(challenge)")
+        do {
+            _ = try self.webauthn.finishAuthentication(
+                credential: input,
+                expectedChallenge: challenge,
+                credentialPublicKey: [UInt8](webAuthnCredential.publicKey.decoded!),
```

Comment:
force unwrap, can we get rid of it?

Reply:
If it fails to decode then we saved bad data into the database, I've changed it to return a server error.